autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java

```diff
@@ -0,0 +1,180 @@
+import com.sequenceiq.periscope.monitor.context.EvaluatorContext;
+import com.sequenceiq.periscope.monitor.evaluator.EvaluatorExecutor;
+import com.sequenceiq.periscope.monitor.evaluator.EventPublisher;
+import com.sequenceiq.periscope.monitor.event.ScalingEvent;
+import com.sequenceiq.periscope.monitor.event.UpdateFailedEvent;
+import com.sequenceiq.periscope.monitor.handler.CloudbreakCommunicator;
+import com.sequenceiq.periscope.repository.LoadAlertRepository;
+import com.sequenceiq.periscope.service.ClusterService;
+import com.sequenceiq.periscope.utils.ClusterUtils;
+import com.sequenceiq.periscope.utils.StackResponseUtils;
+import com.sequenceiq.periscope.utils.TimeUtil;
+
+@Component("YarnLoadEvaluator")
+@Scope("prototype")
+public class YarnLoadEvaluator extends EvaluatorExecutor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(YarnLoadEvaluator.class);
+
+    private static final String EVALUATOR_NAME = YarnLoadEvaluator.class.getName();
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private LoadAlertRepository alertRepository;
+
+    @Inject
+    private EventPublisher eventPublisher;
+
+    @Inject
+    private YarnMetricsClient yarnMetricsClient;
+
+    @Inject
+    private StackResponseUtils stackResponseUtils;
+
+    @Inject
+    private CloudbreakCommunicator cloudbreakCommunicator;
+
+    private long clusterId;
+
+    private Cluster cluster;
+
+    private LoadAlert loadAlert;
+
+    @Nonnull
+    @Override
+    public EvaluatorContext getContext() {
+        return new ClusterIdEvaluatorContext(clusterId);
+    }
+
+    @Override
+    public void setContext(EvaluatorContext context) {
+        clusterId = (long) context.getData();
+    }
+
+    @Override
+    public String getName() {
+        return EVALUATOR_NAME;
+    }
+
+    @Override
+    protected void execute() {
+        cluster = clusterService.findById(clusterId);
+        loadAlert = cluster.getLoadAlerts().stream().findFirst().get();
+
+        long remainingTime = ClusterUtils.getRemainingCooldownTime(
+                loadAlert.getLoadAlertConfiguration().getCoolDownMillis(),
+                cluster.getLastScalingActivity());
+
+        if (remainingTime <= 0) {
+            pollYarnMetricsAndScaleCluster();
+        } else {
+            LOGGER.debug("Cluster cannot be scaled for {} min(s)",
+                    ClusterUtils.TIME_FORMAT.format((double) remainingTime / TimeUtil.MIN_IN_MS));
+        }
+    }
+
+    protected void pollYarnMetricsAndScaleCluster() {
+        long start = System.currentTimeMillis();
+        try {
+            MDCBuilder.buildMdcContext(cluster);
+
+            StackV4Response stackV4Response = cloudbreakCommunicator.getByCrn(cluster.getStackCrn());
+
+            String hostGroupInstanceType =
+                    stackResponseUtils.getHostGroupInstanceType(stackV4Response, loadAlert.getScalingPolicy().getHostGroup());
+
```

Comment:
Check min/max node count constraints first? If this gets reconfigured - then how the YARN response is processed may need to be re-evaluated.
e.g. cluster at 20 nodes. Max reconfigured to 10. YARN replies with 'need more resources'.

Looks like scaleUp handles this - except for bringing the node count down to the maxValue.

Could be a follow up jira as well, since you have already tested this.

Reply:
yes this is handled in scaleUp which is more real time.

Reply:
Bringing the nodeCount down to a dynamically configured 'maxValue' isn't handled from what I can tell.
YARN could end up recommending adding new nodes. maxNodes has been reduced by the end user. This will trigger scaleUp since YARN has a net positive node recommendation.
scaleUp doesn't trigger a reduction in size.

Reply:
the time difference between retrieving latest load alert from db and triggering the scale up\down is just a few milli-seconds ( time taken by yarn api), I didn't make another db hit to get the latest max configured value since this interval is very small and saves one more db hit.  the 'maxValue' can be reconfigured few ms after triggering periscope scaleUpEvent in YarnLoadEvaluator or few ms after sending request to CB even in that case scaleUp would go through.

However in the next evaluation cycle, if yarn doesnt recommend any decommissionNodes and the policy has maxValue which is lesser than currentHostGroup size, if we trigger down scaling it will be forced downscaling, since yarn hasn't really found unused capacity. I dont think this would be expected behaviour because we are taking away required capacity. It will be downscaled automatically when yarn observes reduced load which i think is inline with the eventually consistent nature of autoscaling. I think this would also ensure less aggressive downscaling.

Reply:
Same as above comment.